PEP 691 Updates (#1)
dstufft/peps: pep-0691.rst

To enable parsing responses with only the standard library, this PEP specifies that
all responses (besides the files themselves, and the HTML responses from
:pep:`503`) should be serialized using `JSON <https://www.json.org/>`_.

To enable zero configuration discovery and to minimize the amount of additional HTTP
requests, this PEP extends :pep:`503` such that all of the API endpoints (other than the
files themselves) will utilize HTTP content negotiation to allow client and server to
select the correct serialization format to serve, i.e. either HTML or JSON.

Versioning
----------

Versioning will adhere to :pep:`629` format (``Major.Minor``), which has defined the
existing HTML responses to be ``1.0``. Since this PEP does not introduce new features
into the API, rather it describes a different serialization format for the existing
features, this PEP does not change the existing ``1.0`` version, and instead just
describes how to serialize that into JSON.

> Simililary to :pep:`629`, the major version number **MUST** be incremented if any

Reviewer: ```suggestion
Similar to :pep:`629`, the major version number **MUST** be incremented if any
```

  guaranteed to be available hash is always included.

  By default, any hash algorithm available via `hashlib
  <https://docs.python.org/3/library/hashlib.html>`_ (specifically any that can
  be passed to ``hashlib.new()`` and do not require additional parameters) can
  be used as a key for the hashes dictionary. At least one secure algorithm from
  ``hashlib.algorithms_guaranteed`` **SHOULD** always be included. At the time
  of this PEP, ``sha256`` specifically is recommended.
- ``requires-python``: An **optional** key that exposes the *Requires-Python*
  metadata field, specified in :pep:`345`. Where this is present, installer tools
  **SHOULD** ignore the download when installing to a Python version that
  doesn't satisfy the requirement.

  Unlike ``data-requires-python`` in :pep:`503`, the ``requires-python`` key does not
  require any special escaping other than anything JSON does naturally.
- ``dist-info-metadata-available``: An **optional** key that indicates
  that metadata for this file is available, via the same location as specified in
  :pep:`658` (``{file_url}.metadata``). Where this is present, it **MUST** be true,

>   or a dictionary mapping a hash name to a hex encoded digest of the metadata hash.

Reviewer 2: Let's add a reference to the hashes key above, noting that the contents are similar to that key but for different content?

- **JSON:** ``application/vnd.pypi.simple.v1+json``
- **HTML:** ``application/vnd.pypi.simple.v1+html``

In addition to the above, a special "meta" version is supported named ``latest``,
whose purpose is to allow clients to request the absolute latest version, without
having to know ahead of time what that version is. It is recommended however,
that clients be explicit about what versions they support.

To support existing clients which expect the existing :pep:`503` API responses to
use the ``text/html`` content type, this PEP further defines ``text/html`` as an alias
for the ``application/vnd.pypi.simple.v1+html`` content type.

Version + Format Selection
--------------------------

Now that there is multiple possible serializations, we need a mechanism to allow
clients to indicate what serialization formats that they're able to understand. In
addition, it would be a benefit if any possible new major version to the API can
be added without disrupting existing clients expecting the previous API version.

To enable this, this PEP standardizes on the use of HTTP's
`Server-Driven Content Negotiation <https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation>`_.

> While this PEP won't fully describe the entirity of server-driven content

Reviewer 2: ```suggestion
While this PEP won't fully describe the entirety of server-driven content
```

          "url": "https://example.com/files/holygrail-1.0-py3-none-any.whl",
          "hashes": {"sha256": "...", "blake2b": "..."},
          "requires-python": ">=3.7",
          "dist-info-metadata": true
        }
      ]
    }

Content-Types
-------------

This PEP proposes that all responses from the Simple API will have a standard

> content type that describes what the response is (a simple api response), what

Reviewer: ```suggestion
content type that describes what the response is (a Simple API response), what
```

      all of the possible responses that could have been chosen.
4. The client interprets the response, handling the different types of responses
   that the server may have responded with.

This PEP does not specify which choices the server makes in regards to handling
a content type that it isn't able to return, and clients **SHOULD** be prepared
to handle all of the possible responses in whatever way makes the most sense for
that client.

However, as there is no standard format for how a ``300 Multiple Choices``
response can be interpreted, this PEP highly discourages servers from utilizing
that option, as clients will have no way to understand and select a different
content-type to request. In addition, it's unlikely that the client *could*
understand a different content type anyways, so at best this response would
likely just be treated the same as a ``406 Not Acceptable`` error.

This PEP **does** require that if the meta version ``latest`` is being used, the
server **MUST** respond with the content type for the actual version that is
contained in the response
(i.e. A ``Accept: application/vnd.pypi.simple.latest+json`` request that returns

> a v1.x response should have a ``Content-Type`` of

Reviewer: ```suggestion
a ``v1.x`` response should have a ``Content-Type`` of
```

of the error conditions that standard server-driven content negotiation typically
has (e.g. ``406 Not Available``, ``303 Multiple Choices``, or selecting a default
type to return).

Endpoint Configuration
^^^^^^^^^^^^^^^^^^^^^^

This option technically is not a special option at all, it is just a natural
consequence of using content negotiation and allowing servers to select which of the
available content types is their default.

If a server is unwilling or unable to implement the server-driven content negotiation,
and would instead rather require users to explicitly configure their client to select
the version they want, then that is a supported configuration.

To enable this, a server should make multiple endpoints (for instance,
``/simple/v1+html/`` and/or ``/simple/v1+json/``) for each version+format that they
wish to support. Under that endpoint, they can host a copy of their repository that
only supports one (or a subset) of the content-types. When a client makes a request
using the ``Accept`` header, the server can ignore it and return the content type
that corresponds to that endpoint.

For clients that wish to require specific configuration, they can keep track of

> which version+format a specific repository url was configured for, and when making

Reviewer: ```suggestion
which version+format a specific repository URL was configured for, and when making
```

:pep:`458` does not specify what the target path should be for the Simple API, but I
believe that TUF requires that the target paths be "file-like", in other words, a
path like ``simple/PROJECT/`` is not acceptable, because it technically points to a
directory.

The saving grace is that the target path does not *have* to actually match the URL
being fetched from the Simple API, and it can just be a sigil that the fetching code
knows how to transform into the actual URL that needs to be fetched. This same thing
can hold true for other aspects of the actual HTTP request, such as the ``Accept``
header.

Ultimately figuring out how to map a directory to a filename is out of scope for this
PEP (but it would be in scope for :pep:`458`), and this PEP defers making a decision
about how exactly to represent this inside of :pep:`458` metadata.

However, it appears that the current WIP branch against pip that attempts to implement
:pep:`458` is using a target path like ``simple/PROJECT/index.html``. This could be
modified to include the API version and serialization format using something like
``simple/PROJECT/vnd.pypi.simple.vN.FORMAT``. So the v1 HTML format would be

> ``simple/PROJECT/vnd.pypi.simple.v1.html`` and the v1 JSON format woould be

Reviewer: ```suggestion
``simple/PROJECT/vnd.pypi.simple.v1.html`` and the v1 JSON format would be
```

- When encountering an ``Accept`` header that does not contain any content types
  that it knows how to work with, should not ever return a ``300 Multiple Choice``
  response, and it should be preferred to return a ``406 Not Acceptable`` response.

  - However, if choosing to use the endpoint configuration, you should prefer to
    return a ``200 OK`` response in the expected content type for that endpoint.

- When selecting an acceptable version, should choose the highest version that
  the client supports, with the most expressive/featureful serialization format,
  taking into account the specificity of the client requests as well as any
  quality priority values they have expressed, and it should only use the
  ``text/html`` content type as a last resort.

It is recommended that clients:

- Support all 3 content types described in this PEP, using server-driven
  content negotiation, for as long as they reasonably can.

- When constructing an ``Accept`` header, include all of the content types
  that you support.

  You should generally *not* include a quality priority value for your content
  types, unless you have implementation specific reasons that you want the

>   server to take into account (for example, if you're using the stdlib html

Reviewer: ```suggestion
  server to take into account (for example, if you're using the standard library HTML
```

negotiation to select the most ideal representation of the data.

Does this mean that static servers are no longer supported?
-----------------------------------------------------------

In short, no, static servers are still (almost) fully supported by this PEP.

The specifics of how they are supported will depend on the static server in
question. For example:

- **S3:** S3 fully supports custom content types, however it does not support
  any form of content negotiation. In order to have a server hosted on S3, you
  would have to use the "Endpoint configuration" style of negotiation, and
  users would have to configure their clients explicitly.

> - **Github Pages:** Github pages does not support custom content types, so the

Reviewer: ```suggestion
- **GitHub Pages:** GitHub pages does not support custom content types, so the
```

Why not add an ``application/json`` alias like ``text/html``?
-------------------------------------------------------------

This PEP believes that it is best for both clients and servers to be explicit
about the types of the API responses that are being used, and a content type
like ``application/json`` is the exact opposite of explicit.

The existence of the ``text/html`` alias exists as a compromise primarily to
ensure that existing consumers of the API continue to function as they already
do. There is no such expectation of existing clients using the Simple API with
a ``application/json`` content type.

In addition, ``application/json`` has no versioning in it, which means that
if there is ever a ``2.0`` version of the Simple API, we will be forced to make
a decision. Should ``application/json`` preserve backwards compatibility and
continue to be an alias for ``application/vnd.pypi.simple.v1+json``, or should
it be updated to be an alias for ``application/vnd.pypi.simple.v2+json``?

This problem doesn't exist for ``text/html``, because the assumption is that
HTML will remain a legacy format, and will likely not gain *any* new features,

> much less features that require breaking compatability. So having it be an

Reviewer: ```suggestion
much less features that require breaking compatibility. So having it be an
```

alias for ``application/vnd.pypi.simple.v1+html`` is effectively the same as
having it be an alias for ``application/vnd.pypi.simple.latest+html``, since
``1.0`` will likely be the only HTML version to exist.

The largest benefit to adding the ``application/json`` content type is that
there do things that do not allow you to have custom content types, and require
you to select one of their preset content types. The main example of this being

> Github Pages, which the lack of ``application/json`` support in this PEP means
> that static repositories will no longer be able to be hosted on Github Pages

Reviewer: ```suggestion
GitHub Pages, which the lack of ``application/json`` support in this PEP means
that static repositories will no longer be able to be hosted on GitHub Pages
```
